pkg/execution/executor/executor.go

```diff
@@ -1451,9 +1452,15 @@ func (e *executor) handlePausesAllNaively(ctx context.Context, iter state.PauseI
 			// time to process.  We handle that here and assume that the event
 			// did not occur in time.
 			if pause.Expires.Time().Before(time.Now()) {
-				// Consume this pause to remove it entirely
-				l.Debug().Msg("deleting expired pause")
-				_ = e.pm.DeletePause(context.Background(), *pause)
+				l.Debug().Msg("encountered expired pause")
+
+				shouldDelete := pause.Expires.Time().Add(PauseExpiredDeletionGracePeriod).Before(time.Now())
```

Comment:
This should never happen but it is possible that
- the pause expires without receiving a matching event
- the timeout job has not run yet
- another matching event comes in _after expiry_ and enters this branch, then drops the pause before the timeout job can run, leading to a stuck function run